lib/lint.ml

```diff
@@ -8,459 +8,40 @@ let () = Lwt_preemptive.init 0 1 (fun _errlog -> ()) (* NOTE: Lwt_preemptive is
-type prefix_conflict_class_mismatch =
-  | WrongPrefix of {
-      conflict_class: string;
-      required_prefix: string;
-    }
-  | WrongConflictClass of {
-      prefix: string;
-      required_conflict_class: string;
-    }
-
-type error =
-  | UnnecessaryField of string
-  | UnmatchedName of OpamPackage.Name.t
-  | UnmatchedVersion of OpamPackage.Version.t
-  | DubiousDuneSubst
-  | DuneProjectMissing
-  | DuneConstraintMissing
-  | DuneIsBuild
-  | BadDuneConstraint of string * string
-  | UnexpectedFile of string
-  | ForbiddenPerm of string
-  | OpamLint of (int * [`Warning | `Error] * string)
-  | FailedToDownload of string
-  | NameCollision of string
-  | WeakChecksum of string
-  | PinDepends
-  | ExtraFiles
-  | RestrictedPrefix of string
-  | PrefixConflictClassMismatch of prefix_conflict_class_mismatch
-
-type host_os = Macos | Other [@@deriving to_yojson]
-
 module Check = struct
   type t = unit
 
   let marshal () = Yojson.Safe.to_string `Null
   let unmarshal _ = ()
 
-  let path_from_pkg pkg =
-    "packages" //
-    (OpamPackage.Name.to_string (OpamPackage.name pkg)) //
-    (OpamPackage.to_string pkg)
-
-  let get_opam ~cwd pkg =
-    Analyse.Analysis.get_opam ~cwd (path_from_pkg pkg // "opam") >>/= fun opam ->
-    Lwt.return (OpamFile.OPAM.read_from_string opam)
-
-  let is_perm_correct file =
-    Lwt_unix.stat file >|= function
-    | {st_kind = S_REG; st_perm = 0o644; _} -> true
-    | {st_kind = S_REG; st_perm = 0o664; _} -> true
-    | _ -> false
-
-  let get_files dirname =
-    Lwt_unix.opendir dirname >>= fun dir ->
-    let rec aux files =
-      Lwt.catch begin fun () ->
-        Lwt_unix.readdir dir >>= fun file ->
-        if Fpath.is_rel_seg file then
-          aux files
-        else
-          aux (file :: files)
-      end begin function
-      | End_of_file -> Lwt.return files
-      | exn -> Lwt.fail exn
-      end
-    in
-    Lwt.finalize
-      (fun () -> aux [])
-      (fun () -> Lwt_unix.closedir dir)
-
-  let get_packages ~cwd =
-    get_files @@ Fpath.to_string cwd // "packages"
-
-  let scan_dir ~cwd errors pkg =
-    let dir = Fpath.to_string cwd // path_from_pkg pkg in
-    get_files dir >>= fun files ->
-    let rec aux errors  = function
-      | [] -> Lwt.return errors
-      | "opam"::files ->
-          is_perm_correct (dir // "opam") >|= begin function
-          | true -> errors
-          | false -> ((pkg, ForbiddenPerm (dir // "opam")) :: errors)
-          end >>= fun errors ->
-          aux errors files
-      | file::files ->
-          aux ((pkg, UnexpectedFile file) :: errors) files
-    in
-    aux errors files
-
-  let get_dune_project_version ~pkg url =
-    Lwt_io.with_temp_dir @@ fun dir ->
-    Lwt_preemptive.detach begin fun () ->
-      OpamProcess.Job.run @@
-      OpamRepository.pull_tree (OpamPackage.to_string pkg)
-        (OpamFilename.Dir.of_string dir)
-        (OpamFile.URL.checksum url)
-        [OpamFile.URL.url url]
-    end () >>= function
```

Comment:
Great that we can remove this now! IIUC, this was needed because we were doing long-running operations in the processes by calling the opam libs. But now we are shelling that out, which means the program will not block on completion of the process, but can context switch while we wait for inputs.

We should also remove initialization of `Lwt_preemptive` on line 6 above.

Reply:
Thanks for catching this!